cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/KafkaAdminTopicConfigProvider.java

```diff
@@ -0,0 +1,140 @@
+
+  private Properties _clusterConfigs;
+  private AdminClient _adminClient;
+
+  @Override
+  public Properties clusterConfigs() {
+    return _clusterConfigs;
+  }
+
+  @Override
+  public Properties topicConfigs(String topic) {
+    Config topicConfig = null;
+    ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+    try {
+      LOG.debug("Requesting details for topic '{}'", topic);
+      topicConfig = _adminClient
+              .describeConfigs(Collections.singletonList(topicResource))
+              .all()
+              .get()
+              .get(topicResource);
+    } catch (InterruptedException | ExecutionException e) {
+        LOG.warn("Config check for topic {} failed due to failure to describe its configs.", topic, e);
+    }
+
+    if (topicConfig != null) {
+      return convertTopicConfigToProperties(topicConfig);
+    } else {
+      LOG.error("The configuration for topic '{}' could not be retrieved", topic);
+      return new Properties();
+    }
+  }
+
+  @Override
+  public Map<String, Properties> allTopicConfigs() {
+
+    // Request a map of futures for the config of each topic on the Kafka cluster
+    Map<ConfigResource, KafkaFuture<Config>> topicConfigs = null;
+    try {
+      LOG.debug("Requesting configurations for all topics");
+      topicConfigs = _adminClient
+              .listTopics()
+              .names()
+              .thenApply(
+                      topicNameSet -> _adminClient.describeConfigs(
+                              topicNameSet.stream().map(name -> new ConfigResource(ConfigResource.Type.TOPIC, name)).collect(Collectors.toList())
+                      ).values()
+              )
+              .get();
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.warn("Config check for all topics failed due to failure to describe their configs.", e);
+    }
+
+    Map<String, Properties> propsMap = new HashMap<>();
+    if (topicConfigs != null) {
+
+      // Set a method to run when each topic config future completes which either logs any error or adds the config to the properties map
+      for (Map.Entry<ConfigResource, KafkaFuture<Config>> entry : topicConfigs.entrySet()) {
+
+        entry.getValue().whenComplete((config, error) -> {
+            if (error != null) {
+              LOG.warn("Topic configurations for topic '{}' on the cluster could not be retrieved due to: {}", entry.getKey(), error.getMessage());
+            } else {
+              propsMap.put(entry.getKey().name(), convertTopicConfigToProperties(config));
+            }
+        });
+
+      }
+
+      //Block on all the config futures completing
+      try {
+        KafkaFuture.allOf(topicConfigs.values().toArray(new KafkaFuture[0])).get();
+      } catch (InterruptedException | ExecutionException e) {
+        LOG.warn("Config check for all topics failed due to failure to describe their configs.", e);
+      }
+    }
```

Comment:
I feel we might be able to simplify this logic a bit using a single loop that iterates over `topicConfigs.entrySet()` -- similar to [TopicMinIsrCache#putTopicMinIsr(DescribeConfigsResult)](https://github.com/linkedin/cruise-control/blob/master/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/TopicMinIsrCache.java#L101 ).

Reply:
I have simplified this method in line with the method you suggested. Let me know what you think?